design-document.md

```diff
@@ -232,6 +232,352 @@ Note: Two non-colluding aggregators --- one leader and one standard aggregator
+[HPKE](https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/) public key and
+encryption parameters:
+
+```
+struct {
+  uint8 id; // Used by helper to decide if it recognizes this config.
+  HpkeKemId kem_id;
+  HpkeKdfId kdf_id;
+  HpkePublicKey public_key;
+} HPKEKeyConfig;
+
+opaque HpkePublicKey<0..2^16-1>;
+uint16 HpkeAeadId; // Defined in I-D.irtf-cfrg-hpke.
+uint16 HpkeKemId;  // Defined in I-D.irtf-cfrg-hpke.
+uint16 HpkeKdfId;  // Defined in I-D.irtf-cfrg-hpke.
+```
+
+We call this the helper's *key configation*. The key configuration is used to
+set up a base-mode HPKE context to use to protect the shares sent to the helper
+via the leader.
+
+Finally, the client collects the set of helpers it will upload shares to. It
+ignores a helper if:
+* the client and helper failed to establish a secure, helper-authenticated
+  channel;
+* the request to the helper URL failed or didn't return a valid key config; or
+* the key config specifies a KEM or KDF algorithm the client doesn't recognize.
+
+If the set of helpers is empty, then the client aborts and alerts the leader
+with "no supported helpers".
+
+### Upload
+
+For each helper, the client issues an HTTP request to the leader.
+
+**Request.**
+The client begins by setting up an HPKE context for the helper by running
+
+```
+helper_enc, context = SetupBaseS(pk, PAClientParam.task)
+```
+
+where `pk` is the KEM public key corresponding to `HPKEKeyConfig.public_key`.
+The outputs are the encapsulated context `helper_enc` and the context `context`.
+
+Next, the client encodes its measurements as an input to the given PA protocol.
+It then generates a validity proof for its input and uses `context` to split the
+input and proof into a leader share and a helper share, where the latter is
+encrypted by the HPKE context. (Note that the secret-sharing and encryption
+procedures are specific to the PA protocol.)
+
+[[OPEN ISSUE: Is it safe to generate the proof once, then secret-share between
+each (leader, helper) pair? Probably not in general, but maybe for Prio?]]
+
+Finally, the clients sends an HTTP request to `[leader]/upload` with the
+following payload:
+
+```
+struct {
+  PATask task;                // PAParameter.task
+  uint8 helper_key_config_id; // HPKEKeyConfig.id
+  Url helper_url;             // URL of helper endpoint
+  PAHelperShare helper_share; // The helper's encrypted share
+  PALeaderShare leader_share; // The leader's share
+} PAUplaod;
+
+struct {
+  opaque enc<0..2^16-1>;     // Encapsulated HPKE context
+  opaque payload<0..2^24-1>; // The encrypted share
+} PAHelperShare;
+
+opaque PALeaderShare<0..2^24-1>;
+```
+
+### Input validation
+
+The leader handles HTTP requests to `[leader]/upload` as follows. It first
+checks the request for a well-formed `PAUpload` message. If not found, then the
+leader aborts and alerts the client with "malformed upload". Otherwise, if a
+well-formed `PAUpload` message is found, the leader responds to the request with
+"200 OK".
```

Comment:
Should it wait to send "200 OK" until the helper confirms that the upload is well formed? (That is, it can decrypt the share?)

Reply:
Hmm, that's an interesting suggestion. I'm leaning towards "no" because the leader is permitted to batch a bunch of verify-start requests. This might cause latency that the client would notice. Besides, there's no reason to wait, since the client doesn't care if its shares were deemed valid.